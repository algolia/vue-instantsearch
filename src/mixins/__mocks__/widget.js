let state = {};

// we need to have state given by `component` before it is mounted, otherwise
// we can't render it in most cases (items, hits, etc. are used in the template)
// so we share a (mock) global state during a whole test.
//
// (a mock is imported once per test file, so the state is isolated between tests)
//
// This allows us to import this `__setState` function and call it in the test
// to give the necessary data before mounting.
export function __setState(newState) {
  state = newState;
}

export const createWidgetMixin = () => ({
  data() {
    return {
      state,
    };
  },
});
