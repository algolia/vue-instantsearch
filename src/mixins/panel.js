import Vue from 'vue';

export const PANEL_EMITTER_NAMESPACE = 'instantSearchPanelEmitter';
export const PANEL_CHANGE_EVENT = 'PANEL_CHANGE_EVENT';

export const createPanelProviderMixin = () => ({
  props: {
    emitter: {
      type: Object,
      required: false,
      default() {
        return new Vue({
          name: 'PanelProvider',
        });
      },
    },
  },
  provide() {
    return {
      [PANEL_EMITTER_NAMESPACE]: this.emitter,
    };
  },
  data() {
    return {
      canRefine: true,
    };
  },
  created() {
    this.emitter.$on(PANEL_CHANGE_EVENT, value => {
      this.updateCanRefine(value);
    });
  },
  beforeDestroy() {
    this.emitter.$destroy();
  },
  methods: {
    updateCanRefine(value) {
      this.canRefine = value;
    },
  },
});

export const createPanelConsumerMixin = ({ mapStateToCanRefine }) => ({
  inject: {
    emitter: {
      from: PANEL_EMITTER_NAMESPACE,
      default() {
        return {
          $emit: () => {},
        };
      },
    },
  },
  data() {
    return {
      state: null,
      hasAlreadyEmitted: false,
    };
  },
  watch: {
    state: {
      immediate: true,
      handler(nextState, previousState) {
        if (!nextState) {
          return;
        }

        const previousCanRefine = mapStateToCanRefine(previousState || {});
        const nextCanRefine = mapStateToCanRefine(nextState);

        if (!this.hasAlreadyEmitted || previousCanRefine !== nextCanRefine) {
          this.emitter.$emit(PANEL_CHANGE_EVENT, nextCanRefine);

          this.hasAlreadyEmitted = true;
        }
      },
    },
  },
});
