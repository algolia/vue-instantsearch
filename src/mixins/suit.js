import suit from '../util/suit';

export const createSuitMixin = ({ name }) => ({
  props: {
    classNames: {
      type: Object,
      default: undefined,
    },
  },
  methods: {
    suit(element, modifier) {
      const className = suit(name, element, modifier);
      const userClassName = this.classNames && this.classNames[className];
      if (userClassName) {
        return [className, userClassName].join(' ');
      }
      return className;
    },
  },
});
