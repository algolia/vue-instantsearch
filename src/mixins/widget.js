const defaultShouldUpdate = (previous, next) => previous !== next;

export const createWidgetMixin = ({ connect }) => ({
  inject: {
    instantSearchInstance: {
      name: 'instantSearchInstance',
      default() {
        const tag = this.$options._componentTag;
        throw new TypeError(
          `It looks like you forgot to wrap your Algolia search component
          "<${tag}>" inside of an "<ais-index>" component.`
        );
      },
    },
  },
  data() {
    return {
      state: null,
    };
  },
  computed: {
    widgetParams() {
      const { widgetParams } = this.$options;

      return Object.keys(widgetParams).reduce(
        (acc, name) => ({
          ...acc,
          [name]: widgetParams[name].getValue(this),
        }),
        {}
      );
    },
  },
  watch: {
    widgetParams(next, previous) {
      const { widgetParams } = this.$options;

      const shouldUpdateWidget = Object.keys(widgetParams).some(name => {
        const { shouldUpdate = defaultShouldUpdate } = widgetParams[name];

        return shouldUpdate(previous[name], next[name]);
      });

      if (shouldUpdateWidget) {
        this.instantSearchInstance.removeWidget(this.widget);
        this.widget = this.widgetFactory(next);
        this.instantSearchInstance.addWidget(this.widget);
      }
    },
  },
  created() {
    this.widgetFactory = connect(this.updateData, () => {});
    this.widget = this.widgetFactory(this.widgetParams);
    this.instantSearchInstance.addWidget(this.widget);
  },
  beforeDestroy() {
    this.instantSearchInstance.removeWidget(this.widget);
  },
  methods: {
    updateData(state = {}) {
      this.state = state;
    },
  },
});
