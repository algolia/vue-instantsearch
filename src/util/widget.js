export const addWidget = (connector, parent, widgetParams, render) => {
  const factory = connector(render);
  const widget = factory(widgetParams);
  parent.addWidgets([widget]);

  return widget;
};

export const forceRender = (widget, parent, instantSearchInstance) => {
  if (
    instantSearchInstance.__initialSearchResults &&
    !instantSearchInstance.started
  ) {
    if (typeof instantSearchInstance.__forceRender !== 'function') {
      throw new Error(
        'You are using server side rendering with <ais-instant-search> instead of <ais-instant-search-ssr>.'
      );
    }
    instantSearchInstance.__forceRender(widget, parent);
  }
};

export const removeWidget = (widget, parent) => {
  if (widget) {
    parent.removeWidgets([widget]);
  }
};
