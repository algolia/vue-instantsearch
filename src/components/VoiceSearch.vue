<template>
  <div v-if="state" :class="suit()">
    <slot>
      <button
        type="button"
        :class="suit('button')"
        :title="isBrowserSupported ? buttonTitle : disabledButtonTitle"
        :disabled="!isBrowserSupported"
        @click="handleClick"
      >
        <slot name="buttonText" v-bind="innerSlotProps">
          <svg v-bind="buttonSvgAttrs" v-if="errorNotAllowed">
            <line x1="1" y1="1" x2="23" y2="23"></line>
            <path d="M9 9v3a3 3 0 0 0 5.12 2.12M15 9.34V4a3 3 0 0 0-5.94-.6"></path>
            <path d="M17 16.95A7 7 0 0 1 5 12v-2m14 0v2a7 7 0 0 1-.11 1.23"></path>
            <line x1="12" y1="19" x2="12" y2="23"></line>
            <line x1="8" y1="23" x2="16" y2="23"></line>
          </svg>
          <svg v-bind="buttonSvgAttrs" v-else>
            <path
              d="M12 1a3 3 0 0 0-3 3v8a3 3 0 0 0 6 0V4a3 3 0 0 0-3-3z"
              :fill="isListening ? 'currentColor' : 'none'"
            ></path>
            <path d="M19 10v2a7 7 0 0 1-14 0v-2"></path>
            <line x1="12" y1="19" x2="12" y2="23"></line>
            <line x1="8" y1="23" x2="16" y2="23"></line>
          </svg>
        </slot>
      </button>
      <div :class="suit('status')">
        <slot name="status" v-bind="innerSlotProps">
          <p>{{transcript}}</p>
        </slot>
      </div>
    </slot>
  </div>
</template>

<script>
import connectVoiceSearch from './tmp/connectVoiceSearch';
import { createSuitMixin } from '../mixins/suit';
import { createWidgetMixin } from '../mixins/widget';

export default {
  name: 'AisVoiceSearch',
  mixins: [
    createWidgetMixin({ connector: connectVoiceSearch }),
    createSuitMixin({ name: 'VoiceSearch' }),
  ],
  props: {
    searchAsYouSpeak: {
      type: Boolean,
      required: false,
      default: false,
    },
    buttonTitle: {
      type: String,
      required: false,
      default: 'Search by voice',
    },
    disabledButtonTitle: {
      type: String,
      required: false,
      default: 'Search by voice (not supported on this browser)',
    },
  },
  data() {
    return {
      buttonSvgAttrs: {
        xmlns: 'http://www.w3.org/2000/svg',
        width: '16',
        height: '16',
        viewBox: '0 0 24 24',
        fill: 'none',
        stroke: 'currentColor',
        strokeWidth: '2',
        strokeLinecap: 'round',
        strokeLinejoin: 'round',
      },
    };
  },
  computed: {
    widgetParams() {
      return {
        searchAsYouSpeak: this.searchAsYouSpeak,
      };
    },
    status() {
      return this.state.voiceListeningState.status;
    },
    transcript() {
      return this.state.voiceListeningState.transcript;
    },
    isSpeechFinal() {
      return this.state.voiceListeningState.isSpeechFinal;
    },
    errorCode() {
      return this.state.voiceListeningState.errorCode;
    },
    isBrowserSupported() {
      return this.state.isBrowserSupported;
    },
    isListening() {
      return this.state.isListening;
    },
    errorNotAllowed() {
      return this.status === 'error' && this.errorCode === 'not-allowed';
    },
    innerSlotProps() {
      return {
        status: this.status,
        errorCode: this.errorCode,
        isListening: this.isListening,
        transcript: this.transcript,
        isSpeechFinal: this.isSpeechFinal,
        isBrowserSupported: this.isBrowserSupported,
      };
    },
  },
  methods: {
    toggleListening() {
      this.state.toggleListening();
    },
    handleClick(event) {
      event.currentTarget.blur();
      this.toggleListening();
    },
  },
};
</script>
