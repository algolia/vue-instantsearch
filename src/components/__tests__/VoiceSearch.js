import VoiceSearch from '../VoiceSearch.vue';
import { mount } from '@vue/test-utils';
import { __setState } from '../../mixins/widget';
jest.mock('../../mixins/widget');

const defaultState = {
  voiceListeningState: {
    status: 'initial',
    transcript: undefined,
    isSpeechFinal: undefined,
    errorCode: undefined,
  },
  isBrowserSupported: true,
  isListening: false,
  toggleListening: jest.fn(),
};

const buttonTextScopedSlot = `
  <span slot-scope="{ isListening }">
    {{isListening ? "Stop": "Start"}}
  </span>
`;

describe('button', () => {
  it('calls toggleListening when the button is clicked', () => {
    __setState(defaultState);
    const wrapper = mount(VoiceSearch);
    wrapper.find('button').trigger('click');
    expect(wrapper.vm.state.toggleListening).toHaveBeenCalledTimes(1);
  });
});

describe('Rendering', () => {
  it('renders default template correctly', () => {
    __setState(defaultState);
    const wrapper = mount(VoiceSearch);
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('display the button as disabled on unsupported browser', () => {
    __setState({
      ...defaultState,
      isBrowserSupported: false,
    });
    const wrapper = mount(VoiceSearch);
    expect(wrapper.find('button').attributes().disabled).toBe('disabled');
  });

  it('with custom template for buttonText (1)', () => {
    __setState({
      ...defaultState,
      isListening: true,
    });
    const wrapper = mount(VoiceSearch, {
      scopedSlots: {
        buttonText: buttonTextScopedSlot,
      },
    });
    expect(wrapper.find('button').text()).toBe('Stop');
  });

  it('with custom template for buttonText (2)', () => {
    __setState({
      ...defaultState,
      isListening: false,
    });
    const wrapper = mount(VoiceSearch, {
      scopedSlots: {
        buttonText: buttonTextScopedSlot,
      },
    });
    expect(wrapper.find('button').text()).toBe('Start');
  });

  it('with custom template for status', () => {
    __setState({
      voiceListeningState: {
        status: 'recognizing',
        transcript: 'Hello',
        isSpeechFinal: false,
        errorCode: undefined,
      },
      isBrowserSupported: true,
      isListening: true,
      toggleListening: jest.fn(),
    });
    const wrapper = mount(VoiceSearch, {
      scopedSlots: {
        status: `
          <div slot-scope="{ status, errorCode, isListening, transcript, isSpeechFinal, isBrowserSupported }">
            <p>status: {{status}}</p>
            <p>errorCode: {{errorCode}}</p>
            <p>isListening: {{isListening}}</p>
            <p>transcript: {{transcript}}</p>
            <p>isSpeechFinal: {{isSpeechFinal}}</p>
            <p>isBrowserSupported: {{isBrowserSupported}}</p>
          </div>
        `,
      },
    });
    expect(wrapper.find('.ais-VoiceSearch-status').html()).toMatchSnapshot();
  });
});
