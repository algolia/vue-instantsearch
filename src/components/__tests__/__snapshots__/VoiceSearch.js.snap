// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering renders default template correctly 1`] = `

<div class="ais-VoiceSearch">
  <button type="button"
          title="Search by voice"
          class="ais-VoiceSearch-button"
  >
    <svg xmlns="http://www.w3.org/2000/svg"
         width="16"
         height="16"
         viewbox="0 0 24 24"
         fill="none"
         stroke="currentColor"
         strokewidth="2"
         strokelinecap="round"
         strokelinejoin="round"
    >
      <path d="M12 1a3 3 0 0 0-3 3v8a3 3 0 0 0 6 0V4a3 3 0 0 0-3-3z"
            fill="none"
      >
      </path>
      <path d="M19 10v2a7 7 0 0 1-14 0v-2">
      </path>
      <line x1="12"
            y1="19"
            x2="12"
            y2="23"
      >
      </line>
      <line x1="8"
            y1="23"
            x2="16"
            y2="23"
      >
      </line>
    </svg>
  </button>
  <div class="ais-VoiceSearch-status">
    <p>
    </p>
  </div>
</div>

`;

exports[`Rendering with custom template for status 1`] = `

<div class="ais-VoiceSearch-status">
  <div>
    <p>
      status: recognizing
    </p>
    <p>
      errorCode:
    </p>
    <p>
      isListening: true
    </p>
    <p>
      transcript: Hello
    </p>
    <p>
      isSpeechFinal: false
    </p>
    <p>
      isBrowserSupported: true
    </p>
  </div>
</div>

`;
